pandas/core/algorithms.py

```diff
@@ -929,6 +949,15 @@ def value_counts_internal(
             # Starting in 3.0, we no longer perform dtype inference on the
             #  Index object we construct here, xref GH#56161
             idx = Index(keys, dtype=keys.dtype, name=index_name)
+
+            if (
+                bins is None
+                and not sort
+                and hasattr(values, "freq")
+                and values.freq is not None
```

Comment:
```suggestion
                and hasattr(values, "freq")
                and values.freq is not None
```

Instead could you `isinstance(values, DatetimeIndex)` (you might need a local import of `DatetimeIndex`). Also you can use `values.inferred_freq` to get the frequency instead

Reply:
`DatetimeIndex` is the only relevant object that requires frequency preservation, so it's OK to special case here